### PR TITLE
Enroll CCX coaches in the courses to which they're added

### PR DESCRIPTION
**Pull request by jamiefolsom**
Hi EdX,

First PR for me -- I'm a new member of the MIT ODL engineering team, so I think I'm covered by our institutional agreement. This one is a small change that addresses a rough edge of CCX. 

**Motivation:** MIT ODL issue https://github.com/mitocw/edx-platform/issues/37. A user who is registered but not enrolled in the course, when added as a CCX coach, is not enrolled, and therefore can't see the courseware which s/he is the coach of. In other words, users must be both enrolled manually and added as CCX coaches manually. This PR makes that one step.

**Changes**: Only LMS is affected by this change, and apart from the behavior described below, under "How to test", the only user-visible effect is that the CCX coach gets an email notifying her/him that they've been enrolled in the course. For an instructor adding a coach, there is now one fewer step. This PR makes no change to the behavior of LMS in the case that the user being added is not registered at all.

**How to test**:
[First, enable CCX](http://edx-installing-configuring-and-running.readthedocs.org/en/latest/configuration/enable_ccx.html).

*Before checking out this PR (ie, while running edx-platform master):*

1. Visit the instructor dashboard, (eg: ```/courses/course-v1:edX+DemoX+Demo_Course/instructor#view-membership``)
1. From the "Course Team Management" block, select "CCX Coaches": 
image
1. Enter the username or email of a user who is *not* currently enrolled in the course:
image
1. Note that if you then attempt to unenroll that same user, you get a message, indicating that the user is not enrolled in the course:
image

*After checking out this PR:*

1. Repeat steps 1 and 2 above
1. Remove that user as a CCX coach.
1. Add that user as a CCX coach.
1. Unenroll that user from the course, and you will see that the user's been unenrolled successfully, indicating that the user was enrolled in the process of being added as a CCX coach:
image

Fixes https://github.com/mitocw/edx-platform/issues/37

### CODE REVIEW

**Comment by jamiefolsom** [comment]
A question, probably for you, @pdpinch: I assume that when the user is removed as a coach, we shouldn't automatically unenroll them. Also, this PR doesn't affect what happens if the user doesn't exist yet; ie: it reports not finding that user. Thanks.

**Comment by pdpinch** [comment]
That's correct. Don't unenroll coaches when they are removed from coaching. Most coaches were students first, and we wouldn't know which came first: the student enrollment or the coach role. 

Dealing with users without accounts is out of scope. I'm thinking about a separate PR to invite anyone who is added in any staff role who hasn't registered (b/c this happens *all the time* residentially) but that will require some conversation with edX.

**Comment by jamiefolsom** [comment]
OK, thanks. So I think this PR currently does what the issue requires; if so, the next thing for me to do here is to write a test.

File: lms/djangoapps/instructor/access.py [hunk 1]
```diff
@@ -45,7 +49,7 @@ def allow_access(course, user, level):
     """
     Allow user access to course modification.
 
-    `level` is one of ['instructor', 'staff', 'beta']
+    `level` is one of ['instructor', 'staff', 'beta', 'ccx_coach']
```
Comment (pdpinch) [inline lms/djangoapps/instructor/access.py]: Why did you make this change? I'm not sure what this permission does but we generally don't want coaches to modify the course. 

That said, beta users shouldn't be able to modify the course either...

Comment (jamiefolsom) [inline lms/djangoapps/instructor/access.py]: Good question... appears I was a bit over-zealous in making the various access levels consistent.

**Comment by pdpinch** [comment]
Beside my question about roles, this is looking good. 

- [ ] As I mentioned yesterday, this will need a "cover letter" when it goes to edX. See http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/process/cover-letter.html which is part of http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/process/overview.html

- [ ] Since this is a pretty narrow PR I think you just need to explain the case & maybe how to test it. The important info for the reviewer is that this only affects lms. The only user facing feature is the e-mail that is sent to the Coach. There probably should be an updated to the CCX documentation to reflect this though (that's in a different repo). 

- [ ] You'll also need to tell them you work for MIT ODL and that your contribution is covered by our institutional agreement (so you don't have to sign a contribution agreement)

- [ ] And finally, after the PR is opened on edx/edx-platform, you'll want to open another one to add yourself to the contributors file. It's best to keep that separate, otherwise you'll be in merge-conflict hell.

**Comment by jamiefolsom** [comment]
@pdpinch cool, thanks, will do those things.

**Comment by jamiefolsom** [comment]
Oh, one thing -- would you like the student to get an email? I didn't specify that in the enroll_email call, and the default is false.

**Comment by pdpinch** [comment]
I thought you were just dealing with coaches, not students?

Can you point me to the line of code in question?

File: lms/djangoapps/instructor/access.py [hunk 2]
```diff
@@ -75,6 +79,8 @@ def _change_access(course, user, level, action):
         raise ValueError("unrecognized level '{}'".format(level))
 
     if action == 'allow':
+        if level == 'ccx_coach':
+            before, after, enrollment_obj = enroll_email(course_id=course.id, student_email=user.email, auto_enroll=True)
```
Comment (jamiefolsom) [inline lms/djangoapps/instructor/access.py]: Here's the call to enroll_email. And the method signature for enroll email is:

```def enroll_email(course_id, student_email, auto_enroll=False, email_students=False, email_params=None, language=None):```

Quoting the comments on that method, ````email_students` determines if student should be notified of action by email.```

So by not specifying that parameter, the default is used, and the new coach is not emailed to notify them they've been enrolled.

Comment (pdpinch) [inline lms/djangoapps/instructor/access.py]: I see. 

I'd like to e-mail the coach. You should probably mention that in the cover letter as well. 

- Peter

Comment (jamiefolsom) [inline lms/djangoapps/instructor/access.py]: Sure, will do.

**Comment by jamiefolsom** [comment]
Hi @pdpinch, Emails are now getting sent, and I've drafted a cover letter. Let me know if you think this looks good. Thanks!

**Comment by pdpinch** [comment]
Nice cover letter. Maybe more detail than you need, but they'll appreciate you for it. I feel like the sentence for the motivation needs a period somewhere, or bullets, but it's not a big deal.

**Comment by jamiefolsom** [comment]
@pdpinch eager to get off on the right foot. You were right about the motivation sentence -- I ran out of breath just reading it to myself. Adjusted. Good to submit to edx? Thanks.

**Comment by jamiefolsom** [comment]
Hey @mitocw/odl-engineering can I/should I merge to mitocw/master before opening the edx PR? Thanks.

**Comment by jamiefolsom** [comment]
Per conversation with @noisecapella -- closing this PR in advance of opening one on edx/master.

**Comment by jamiefolsom** [comment]
Odd closing/labelling behavior, which I am guessing is related to waffle.io settings. Can't manually close via github (keeps getting re-opened/labelled as seen above), but moving a PR to Done in waffle.io closes the PR. FYI, @pdpinch.

**Pull request closed without merging**